internal/jujucommands/jujucommands.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2025 Canonical.
+
+// Package jujucommands provides functions run juju cmds from a JIMM instance.
+// Each command function is run with its own isolated in-mem store.
+package jujucommands
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/juju/cmd/v3"
+	"github.com/juju/juju/cmd/juju/commands"
+	"github.com/juju/juju/jujuclient"
+	"github.com/mitchellh/go-linereader"
+)
+
+var (
+	whitelist = []string{"help", "bootstrap"}
+)
+
+type outputLine struct {
+	Line string
+	Err  error
+}
+
+// runCmdWithOutputRetriever runs the command given by cmdAndArgs using the provided jujuclient.ClientStore.
+// It returns a channel that streams OutputLine values for each line of command output and any final error.
+//
+// Upon receiving an error in an OutputLine, the line previously streamed is guaranteed to be the error line written by
+// the command.
+// I.e., if you run: "help -b"
+// You will get:
+//
+//	Output line: ERROR option provided but not defined: -b // The error line from juju (Which is a none-erroneous OutputLine)
+//	Command finished with error: cmd failed with code 2 // The command error code line (Contains an error in the OutputLine)
+//
+// Consumers are expected to simply read from the OutputLine channel and check for an error like so:
```

Comment:
I'd add a section/line to explain that if the consumer is not timely in consuming lines from the output channel, then the command execution will block.

Reply:
I'm not certain it blocks, I'm guessing it does between the cmd I/o is piped and reader would hold it up - but... Without checking I'm not 100% - it could buffer somewhere.